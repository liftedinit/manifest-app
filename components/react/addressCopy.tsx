import React, { useState, useEffect } from 'react';
import { FiCopy, FiCheck } from 'react-icons/fi';

export const TruncatedAddressWithCopy = ({
  address,
  slice,
  size,
}: {
  address: string;
  slice: number;
  size?: string;
}) => {
  const [copied, setCopied] = useState(false);

  useEffect(() => {
    let timer: ReturnType<typeof setTimeout>;
    if (copied) {
      timer = setTimeout(() => setCopied(false), 2000);
    }
    return () => clearTimeout(timer);
  }, [copied]);

  const handleCopy = async (e: React.MouseEvent<HTMLDivElement>) => {
    e.stopPropagation();
    e.preventDefault();
    try {
      await navigator.clipboard.writeText(address);
      setCopied(true);
    } catch (err) {
      console.error('Failed to copy: ', err);
    }
  };

  const truncatedAddress = `${address?.slice(0, slice)}...${address?.slice(-6)}`;
  const iconSize = size === 'small' ? 10 : 16;

  return (
    <span
      className="flex items-center hover:text-primary dark:hover:text-primary space-x-2 text-[#00000099] dark:text-[#FFFFFF99]"
      onClick={handleCopy}
      style={{ cursor: 'pointer' }}
    >
      <span className="truncate ">{truncatedAddress}</span>
      {copied ? <FiCheck size={iconSize} /> : <FiCopy size={iconSize} />}
    </span>
  );
};

export const AddressWithCopy = ({ address }: { address: string }) => {
  const [copied, setCopied] = useState(false);

  useEffect(() => {
    let timer: ReturnType<typeof setTimeout>;
    if (copied) {
      timer = setTimeout(() => setCopied(false), 2000);
    }
    return () => clearTimeout(timer);
  }, [copied]);

  const handleCopy = async () => {
    try {
      await navigator.clipboard.writeText(address);
      setCopied(true);
    } catch (err) {
      console.error('Failed to copy: ', err);
    }
  };

  return (
    <span
      className="flex items-center space-x-2"
      onClick={handleCopy}
      style={{ cursor: 'pointer' }}
    >
      <span className="hover:text-primary">{address}</span>
      {copied ? <FiCheck size="16" /> : <FiCopy size="16" />}
    </span>
  );
};
