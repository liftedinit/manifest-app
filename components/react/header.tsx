import env from '@/config/env';
import { WalletSection } from '../wallet';

export default function Header() {
  return (
    <header>
      <nav className="bg-white border-b border-gray-200 dark:bg-gray-800 dark:border-gray-700 px-4 fixed w-full lg:px-6 py-2 ">
        <div className="flex justify-end items-center mx-auto max-w-screen-6xl">
          {/* Log Out Button */}
          <WalletSection chainName={env.chain} />
        </div>
      </nav>
    </header>
  );
}
