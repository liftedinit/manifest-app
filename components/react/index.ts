export * from './addressCopy';
export * from './modal';
export * from './views';
export * from './settingsModal';
export * from './inputs';
export { SignModal } from './authSignerModal';
export * from './ModalDialog';
