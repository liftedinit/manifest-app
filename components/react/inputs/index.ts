export * from './AmountInput';
export * from './TextInput';
export * from './NumberInput';
export * from './TextArea';
