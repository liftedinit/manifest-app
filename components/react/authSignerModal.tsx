import { useEffect, useState } from 'react';
import { SignData } from '@cosmos-kit/web3auth';
import { TxBody, AuthInfo } from '@liftedinit/manifestjs/dist/codegen/cosmos/tx/v1beta1/tx';
import { decodePubkey } from '@cosmjs/proto-signing';
import { useWallet, useChain } from '@cosmos-kit/react';
import { getRealLogo } from '@/utils';
import { useTheme } from '@/contexts';
import env from '@/config/env';
import { ArrowRightIcon } from '../icons';
import { objectSyntax } from '../messageSyntax';
import { MsgSend } from '@liftedinit/manifestjs/dist/codegen/cosmos/bank/v1beta1/tx';
import {
  MsgCreateGroupWithPolicy,
  MsgSubmitProposal,
  MsgUpdateGroupMembers,
  MsgUpdateGroupPolicyMetadata,
  MsgUpdateGroupPolicyDecisionPolicy,
  MsgUpdateGroupMetadata,
} from '@liftedinit/manifestjs/dist/codegen/cosmos/group/v1/tx';
import {
  MsgCancelUpgrade,
  MsgSoftwareUpgrade,
} from '@liftedinit/manifestjs/dist/codegen/cosmos/upgrade/v1beta1/tx';
import { MsgSetPower } from '@liftedinit/manifestjs/dist/codegen/strangelove_ventures/poa/v1/tx';
import {
  MsgPayout,
  MsgBurnHeldBalance,
} from '@liftedinit/manifestjs/dist/codegen/liftedinit/manifest/v1/tx';
import {
  MsgSetDenomMetadata,
  MsgCreateDenom,
} from '@liftedinit/manifestjs/dist/codegen/osmosis/tokenfactory/v1beta1/tx';

type DisplayDataToSignProps = {
  data: SignData;
  address: string;
};

// Message decoder registry
const messageDecoders: Record<string, (value: Uint8Array) => any> = {
  '/cosmos.bank.v1beta1.MsgSend': (value: Uint8Array) => {
    const decoded = MsgSend.decode(value);
    return { ...decoded };
  },
  '/cosmos.group.v1.MsgCreateGroupWithPolicy': (value: Uint8Array) => {
    const decoded = MsgCreateGroupWithPolicy.decode(value);
    return { ...decoded };
  },
  '/cosmos.group.v1.MsgSubmitProposal': (value: Uint8Array) => {
    const decoded = MsgSubmitProposal.decode(value);
    return { ...decoded };
  },
  '/cosmos.group.v1.MsgUpdateGroupMetadata': (value: Uint8Array) => {
    const decoded = MsgUpdateGroupMetadata.decode(value);
    return { ...decoded };
  },
  '/cosmos.group.v1.MsgUpdateGroupPolicyMetadata': (value: Uint8Array) => {
    const decoded = MsgUpdateGroupPolicyMetadata.decode(value);
    return { ...decoded };
  },
  '/cosmos.group.v1.MsgUpdateGroupPolicyDecisionPolicy': (value: Uint8Array) => {
    const decoded = MsgUpdateGroupPolicyDecisionPolicy.decode(value);
    return { ...decoded };
  },
  '/cosmos.group.v1.MsgUpdateGroupMembers': (value: Uint8Array) => {
    const decoded = MsgUpdateGroupMembers.decode(value);
    return { ...decoded };
  },
  '/cosmos.upgrade.v1beta1.MsgCancelUpgrade': (value: Uint8Array) => {
    const decoded = MsgCancelUpgrade.decode(value);
    return { ...decoded };
  },
  '/cosmos.upgrade.v1beta1.MsgSoftwareUpgrade': (value: Uint8Array) => {
    const decoded = MsgSoftwareUpgrade.decode(value);
    return { ...decoded };
  },
  '/strangelove_ventures.poa.v1.MsgSetPower': (value: Uint8Array) => {
    const decoded = MsgSetPower.decode(value);
    return { ...decoded };
  },
  '/liftedinit.manifest.v1.MsgPayout': (value: Uint8Array) => {
    const decoded = MsgPayout.decode(value);
    return { ...decoded };
  },
  '/liftedinit.manifest.v1.MsgBurnHeldBalance': (value: Uint8Array) => {
    const decoded = MsgBurnHeldBalance.decode(value);
    return { ...decoded };
  },
  '/osmosis.tokenfactory.v1beta1.MsgSetDenomMetadata': (value: Uint8Array) => {
    const decoded = MsgSetDenomMetadata.decode(value);
    return { ...decoded };
  },
  '/osmosis.tokenfactory.v1beta1.MsgCreateDenom': (value: Uint8Array) => {
    const decoded = MsgCreateDenom.decode(value);
    return { ...decoded };
  },
};

const DisplayDataToSign = ({
  data,
  address,
  className,
  addressClassName,
  txInfoClassName,
  theme,
}: DisplayDataToSignProps & {
  className?: string;
  addressClassName?: string;
  txInfoClassName?: string;
  theme?: string;
}) => {
  const [isAddressExpanded, setIsAddressExpanded] = useState(false);
  const [isTxInfoExpanded, setIsTxInfoExpanded] = useState(false);

  const decodeBodyBytes = (bodyBytes: Uint8Array) => {
    try {
      const decodedBody = TxBody.decode(bodyBytes);
      return {
        messages: decodedBody.messages.map(msg => {
          const base64Value = Buffer.from(msg.value).toString('base64');

          try {
            // Check if we have a specific decoder for this message type
            if (messageDecoders[msg.typeUrl]) {
              return {
                typeUrl: msg.typeUrl,
                value: messageDecoders[msg.typeUrl](msg.value),
              };
            }

            // Fallback to generic base64 decoding
            const decodedValue = Buffer.from(base64Value, 'base64').toString('utf8');
            try {
              return {
                typeUrl: msg.typeUrl,
                value: JSON.parse(decodedValue),
              };
            } catch {
              return {
                typeUrl: msg.typeUrl,
                value: decodedValue,
              };
            }
          } catch (error) {
            console.error(`Failed to decode message of type ${msg.typeUrl}:`, error);
            return {
              typeUrl: msg.typeUrl,
              value: base64Value,
            };
          }
        }),
        memo: decodedBody.memo,
        timeoutHeight: decodedBody.timeoutHeight.toString(),
        extensionOptions: decodedBody.extensionOptions,
        nonCriticalExtensionOptions: decodedBody.nonCriticalExtensionOptions,
      };
    } catch (error) {
      console.error('Failed to decode bodyBytes:', error);
      return 'Failed to decode bodyBytes';
    }
  };

  const decodeAuthInfoBytes = (authInfoBytes: Uint8Array) => {
    try {
      const decodedAuthInfo = AuthInfo.decode(authInfoBytes);
      return {
        signerInfos: decodedAuthInfo.signerInfos.map(signerInfo => ({
          publicKey: signerInfo.publicKey ? decodePubkey(signerInfo.publicKey) : null,
          modeInfo: signerInfo.modeInfo,
          sequence: signerInfo.sequence.toString(),
        })),
        fee: {
          amount: decodedAuthInfo.fee?.amount,
          gasLimit: decodedAuthInfo.fee?.gasLimit.toString(),
          payer: decodedAuthInfo.fee?.payer,
          granter: decodedAuthInfo.fee?.granter,
        },
      };
    } catch (error) {
      console.error('Failed to decode authInfoBytes:', error);
      return 'Failed to decode authInfoBytes';
    }
  };

  const formatValue = (value: any, theme: string) => {
    if (value instanceof Uint8Array) {
      return Buffer.from(value).toString('base64');
    }
    if (typeof value === 'object' && value !== null) {
      if ('bodyBytes' in value && 'authInfoBytes' in value) {
        const decodedValue = {
          ...value,
          bodyBytes: decodeBodyBytes(value.bodyBytes),
          authInfoBytes: decodeAuthInfoBytes(value.authInfoBytes),
        };
        return objectSyntax(
          JSON.parse(
            JSON.stringify(decodedValue, (_, v) => (typeof v === 'bigint' ? v.toString() : v))
          ),
          theme
        );
      }
      return objectSyntax(
        JSON.parse(JSON.stringify(value, (_, v) => (typeof v === 'bigint' ? v.toString() : v))),
        theme
      );
    }
    return String(value);
  };

  return (
    <div className={className}>
      <div className="flex flex-col gap-2">
        <button
          onClick={() => setIsAddressExpanded(!isAddressExpanded)}
          className="flex items-center gap-2 text-sm font-medium"
        >
          <div className="flex items-center gap-2 flex-row justify-between">
            <span>Address</span>
            {isAddressExpanded ? (
              <ArrowRightIcon size={12} style={{ transform: 'rotate(90deg)' }} />
            ) : (
              <ArrowRightIcon size={12} style={{ transform: 'rotate(180deg)' }} />
            )}
          </div>
        </button>
        {isAddressExpanded && <pre className={addressClassName}>{address}</pre>}
      </div>
      <div className="flex flex-col gap-2">
        <button
          onClick={() => setIsTxInfoExpanded(!isTxInfoExpanded)}
          className="flex items-center gap-2 text-sm font-medium"
        >
          <div className="flex items-center gap-2 flex-row justify-between">
            <span>Tx Info</span>
            {isTxInfoExpanded ? (
              <ArrowRightIcon size={12} style={{ transform: 'rotate(90deg)' }} />
            ) : (
              <ArrowRightIcon size={12} style={{ transform: 'rotate(180deg)' }} />
            )}
          </div>
        </button>
        {isTxInfoExpanded && (
          <div className={txInfoClassName}>{formatValue(data.value, theme ?? 'light')}</div>
        )}
      </div>
    </div>
  );
};

const SignModal = ({
  visible,
  onClose,
  data,
  approve,
  reject,
}: {
  visible: boolean;
  onClose: () => void;
  data: SignData;
  approve: () => void;
  reject: () => void;
}) => {
  const wallet = useWallet();
  const { address } = useChain(env.chain);
  const walletIcon = wallet.wallet?.logo;
  const walletName = wallet.wallet?.prettyName;
  const { theme } = useTheme();

  useEffect(() => {
    const modal = document.getElementById('sign-modal') as HTMLDialogElement;
    if (visible) {
      modal.showModal();
    } else {
      modal.close();
    }
  }, [visible]);

  const walletIconString = walletIcon?.toString() ?? '';

  return (
    <dialog id="sign-modal" className="modal top-0 right-0">
      <div className="modal-box max-w-lg w-full dark:bg-[#1D192D] bg-[#FFFFFF] rounded-lg shadow-xl">
        <div className="flex justify-between items-center pb-4">
          <div className="flex items-center gap-3">
            <img
              src={getRealLogo(walletIconString, theme === 'dark')}
              alt="Wallet type logo"
              className="w-8 h-8"
            />
            <h3 className="text-xl font-semibold">{walletName?.toString()} Direct Signer</h3>
          </div>
          <button className="btn btn-sm btn-circle btn-ghost" onClick={onClose}>
            ✕
          </button>
        </div>

        <DisplayDataToSign
          data={data}
          address={address ?? ''}
          theme={theme}
          className="space-y-4"
          addressClassName="p-3 rounded-md text-sm overflow-auto h-12 dark:bg-[#E0E0FF0A] bg-[#E0E0FF0A] dark:border-[#FFFFFF33] border-[#00000033] border"
          txInfoClassName="p-3 rounded-md text-sm overflow-auto h-[32rem] dark:bg-[#E0E0FF0A] bg-[#E0E0FF0A] dark:border-[#FFFFFF33] border-[#00000033] border"
        />

        <div className="modal-action mt-6 flex justify-between gap-4">
          <button
            className="btn flex-1 rounded-[12px] focus:outline-none dark:bg-[#FFFFFF0F] bg-[#0000000A]"
            onClick={() => {
              reject();
              onClose();
            }}
          >
            Reject
          </button>
          <button
            className="btn btn-gradient flex-1 rounded-[12px]"
            onClick={() => {
              approve();
              onClose();
            }}
          >
            Approve
          </button>
        </div>
      </div>
      <form method="dialog" className="modal-backdrop">
        <button onClick={onClose}>close</button>
      </form>
    </dialog>
  );
};

export default SignModal;
