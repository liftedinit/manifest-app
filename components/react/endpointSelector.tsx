import React, { useState } from 'react';
import { useQuery } from '@tanstack/react-query';
import { Formik, Form } from 'formik';
import dynamic from 'next/dynamic';

import { useToast } from '@/contexts';
import { useEndpointStore } from '@/store/endpointStore';
import { TextInput } from './inputs';
import Yup from '@/utils/yupExtensions';

export interface Endpoint {
  rpc: string;
  api: string;
  provider: string;
  isHealthy: boolean;
  network: 'mainnet' | 'testnet';
  custom: boolean;
}

const validateRPCEndpoint = async (url: string) => {
  console.log('Validating RPC endpoint:', url);
  if (!url) return false;
  try {
    const endpoint = url.startsWith('http') ? url : `https://${url}`;
    console.log('Making RPC request to:', endpoint);

    const response = await fetch(endpoint, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        jsonrpc: '2.0',
        id: 1,
        method: 'status',
        params: [],
      }),
    });

    console.log('RPC Response status:', response.status);
    if (!response.ok) {
      console.log('RPC Response not ok');
      return false;
    }

    const data = await response.json();
    console.log('RPC Response data:', data);

    return data?.result?.sync_info?.catching_up === false;
  } catch (error) {
    console.error('RPC Validation error:', error);
    return false;
  }
};

const validateAPIEndpoint = async (url: string) => {
  console.log('Validating API endpoint:', url);
  if (!url) return false;
  try {
    const endpoint = url.startsWith('http') ? url : `https://${url}`;
    const baseUrl = endpoint.endsWith('/') ? endpoint.slice(0, -1) : endpoint;
    const apiUrl = `${baseUrl}/cosmos/base/tendermint/v1beta1/syncing`;

    console.log('Making API request to:', apiUrl);

    const response = await fetch(apiUrl);
    console.log('API Response status:', response.status);

    if (!response.ok) {
      console.log('API Response not ok');
      return false;
    }

    return response.ok;
  } catch (error) {
    console.error('API Validation error:', error);
    return false;
  }
};

const validateIndexerEndpoint = async (url: string) => {
  console.log('Validating Indexer endpoint:', url);
  if (!url) return false;
  try {
    const endpoint = url.startsWith('http') ? url : `https://${url}`;
    const baseUrl = endpoint.endsWith('/') ? endpoint.slice(0, -1) : endpoint;
    const indexerUrl = `${baseUrl}`;

    console.log('Making Indexer request to:', indexerUrl);

    const response = await fetch(indexerUrl);
    console.log('Indexer Response status:', response.status);

    if (!response.ok) {
      console.log('Indexer Response not ok');
      return false;
    }

    return response.ok;
  } catch (error) {
    console.error('Indexer Validation error:', error);
    return false;
  }
};

const validateExplorerEndpoint = async (url: string) => {
  console.log('Validating Explorer endpoint:', url);
  if (!url) return false;
  try {
    // const endpoint = url.startsWith('http') ? url : `https://${url}`;
    // const baseUrl = endpoint.endsWith('/') ? endpoint.slice(0, -1) : endpoint;
    // const explorerUrl = `${baseUrl}`;

    // console.log('Making Explorer request to:', explorerUrl);
    //
    // const response = await fetch(explorerUrl);
    // console.log('Explorer Response status:', response.status);
    //
    // if (!response.ok) {
    //   console.log('Explorer Response not ok');
    //   return false;
    // }
    //
    // return response.ok;
    return true;
  } catch (error) {
    console.error('Explorer Validation error:', error);
    return false;
  }
};

const EndpointSchema = Yup.object().shape({
  rpc: Yup.string().required('RPC endpoint is required').test({
    name: 'rpc-validation',
    message: 'RPC endpoint is not responding',
    test: validateRPCEndpoint,
  }),
  api: Yup.string().required('API endpoint is required').test({
    name: 'api-validation',
    message: 'API endpoint is not responding',
    test: validateAPIEndpoint,
  }),
  indexer: Yup.string().required('Indexer endpoint is required').test({
    name: 'indexer-validation',
    message: 'Indexer endpoint is not responding',
    test: validateIndexerEndpoint,
  }),
  explorer: Yup.string().required('Explorer endpoint is required').test({
    name: 'explorer-validation',
    message: 'Explorer endpoint is not responding',
    test: validateExplorerEndpoint,
  }),
});

function SSREndpointSelector() {
  const {
    endpoints,
    selectedEndpointKey,
    setSelectedEndpointKey,
    addEndpoint,
    removeEndpoint,
    updateEndpointHealth,
  } = useEndpointStore(state => ({
    endpoints: state.endpoints,
    selectedEndpointKey: state.selectedEndpointKey,
    setSelectedEndpointKey: state.setSelectedEndpointKey,
    addEndpoint: state.addEndpoint,
    removeEndpoint: state.removeEndpoint,
    updateEndpointHealth: state.updateEndpointHealth,
  }));

  const { setToastMessage } = useToast();
  const [isModalOpen, setIsModalOpen] = useState(false);
  const [endpointToRemove, setEndpointToRemove] = useState<string | null>(null);

  const { isLoading, error } = useQuery({
    queryKey: ['checkEndpoints', endpoints],
    queryFn: updateEndpointHealth,
    refetchInterval: 30000,
    enabled: true,
  });

  const handleCustomEndpointSubmit = async (values: {
    rpc: string;
    api: string;
    indexer: string;
    explorer: string;
  }) => {
    const rpcUrl = values.rpc.startsWith('http') ? values.rpc : `https://${values.rpc}`;
    const apiUrl = values.api.startsWith('http') ? values.api : `https://${values.api}`;
    const indexerUrl = values.indexer.startsWith('http')
      ? values.indexer
      : `https://${values.indexer}`;
    const explorerUrl = values.explorer.startsWith('http')
      ? values.explorer
      : `https://${values.explorer}`;

    try {
      const [isRPCValid, isAPIValid] = await Promise.all([
        validateRPCEndpoint(rpcUrl),
        validateAPIEndpoint(apiUrl),
      ]);

      if (!isRPCValid || !isAPIValid) {
        throw new Error('Endpoint validation failed');
      }

      await addEndpoint(rpcUrl, apiUrl, indexerUrl, explorerUrl);
      setToastMessage({
        type: 'alert-success',
        title: 'Custom endpoint added',
        description: 'The new endpoint has been successfully added.',
        bgColor: '#2ecc71',
      });
    } catch (error) {
      console.error('Error adding custom endpoint:', error);
      let errorMessage = 'An unknown error occurred while adding the endpoint.';

      if (error instanceof Error) {
        if (error.message.includes('Invalid URL')) {
          errorMessage = 'Invalid URL format. Please check all URLs.';
        } else if (error.message.includes('Network error')) {
          errorMessage = 'Network error. Please check your internet connection and try again.';
        } else if (error.message.includes('Timeout')) {
          errorMessage = 'Connection timeout. The endpoint might be unreachable.';
        } else {
          errorMessage = error.message;
        }
      }

      setToastMessage({
        type: 'alert-error',
        title: 'Error adding custom endpoint',
        description: errorMessage,
        bgColor: '#e74c3c',
      });
      throw error;
    }
  };

  const truncateUrl = (url: string) => {
    try {
      const ipRegex = /^(?:\d{1,3}\.){3}\d{1,3}(?::\d+)?$/;
      if (!url.startsWith('http://') && !url.startsWith('https://')) {
        url = ipRegex.test(url) ? `http://${url}` : `https://${url}`;
      }
      const parsedUrl = new URL(url);
      return `${parsedUrl.host}`;
    } catch (error) {
      console.error('Invalid URL:', url);
      return url;
    }
  };

  const isCustomEndpoint = (endpoint: Endpoint) => endpoint.custom;

  return (
    <dialog id="endpoint_selector_modal" className="modal">
      <div className="modal-box w-full max-w-2xl bg-[#FFFFFF] dark:bg-[#1D192D]">
        <h3 className="mb-4 text-xl font-extrabold tracking-tight sm:mb-6 leading-tight border-b-[0.5px] dark:text-[#FFFFFF99] dark:border-[#FFFFFF99] border-b-[black] pb-4">
          Available Endpoints
        </h3>

        <div className="dark:bg-[#FFFFFF0F] bg-[#FFFFFFCC] p-[24px] rounded-[24px] mb-4">
          {isLoading ? (
            <p>Checking endpoints...</p>
          ) : error ? (
            <p>Error checking endpoints</p>
          ) : (
            <ul className="space-y-4">
              {endpoints.map((endpoint: Endpoint, index: number) => (
                <li
                  key={index}
                  className={`
                    flex flex-col p-3 rounded-lg cursor-pointer border
                    ${
                      selectedEndpointKey === endpoint.provider
                        ? 'dark:bg-[#FFFFFF14] bg-[#00000008] dark:border-[#FFFFFF30] border-[#00000030]'
                        : 'dark:border-[#FFFFFF10] border-[#00000015] hover:dark:bg-[#FFFFFF08] hover:bg-[#00000005]'
                    }
                    transition-all duration-200 group
                  `}
                  onClick={() => setSelectedEndpointKey(endpoint.provider)}
                >
                  <div className="flex justify-between items-center gap-4">
                    <div className="flex-1 min-w-0">
                      <div className="flex items-center gap-2">
                        <div
                          className={`w-2 h-2 rounded-full transition-all duration-300
                            ${
                              endpoint.isHealthy
                                ? 'bg-success shadow-sm shadow-success/30'
                                : 'bg-error shadow-sm shadow-error/30'
                            }
                          `}
                        ></div>
                        <p className="text-base font-medium truncate">
                          {endpoint.custom ? `Custom (${endpoint.network})` : endpoint.provider}
                        </p>
                      </div>
                      <p className="text-sm opacity-60 mt-1 truncate">
                        {truncateUrl(endpoint.rpc)}
                      </p>
                    </div>

                    {isCustomEndpoint(endpoint) && (
                      <button
                        className={`
                          btn btn-sm btn-ghost opacity-0 group-hover:opacity-100
                          hover:bg-error/10 hover:text-error
                          transition-all duration-200
                          ${endpointToRemove === endpoint.provider ? 'opacity-100 rotate-90' : ''}
                        `}
                        onClick={e => {
                          e.stopPropagation();
                          if (endpointToRemove === endpoint.provider) {
                            removeEndpoint(endpoint.provider);
                            setEndpointToRemove(null);
                          } else {
                            setEndpointToRemove(endpoint.provider);
                          }
                        }}
                      >
                        <svg
                          xmlns="http://www.w3.org/2000/svg"
                          className="h-4 w-4"
                          fill="none"
                          viewBox="0 0 24 24"
                          stroke="currentColor"
                        >
                          {endpointToRemove === endpoint.provider ? (
                            <path
                              strokeLinecap="round"
                              strokeLinejoin="round"
                              strokeWidth={2}
                              d="M6 18L18 6M6 6l12 12"
                            />
                          ) : (
                            <path
                              strokeLinecap="round"
                              strokeLinejoin="round"
                              strokeWidth={2}
                              d="M6 12h12"
                            />
                          )}
                        </svg>
                      </button>
                    )}
                  </div>
                </li>
              ))}
            </ul>
          )}
        </div>

        <button
          className="btn btn-gradient text-white w-full"
          onClick={e => {
            e.stopPropagation();
            setIsModalOpen(true);
          }}
        >
          Add Custom Endpoints
        </button>
      </div>

      <form method="dialog" className="modal-backdrop">
        <button>close</button>
      </form>

      {isModalOpen && (
        <div className="modal modal-open">
          <div className="modal-box w-full bg-[#FFFFFF] dark:bg-[#1D192D] max-w-2xl">
            <h3 className="mb-4 text-xl font-extrabold tracking-tight sm:mb-6 leading-tight border-b-[0.5px] dark:text-[#FFFFFF99] dark:border-[#FFFFFF99] border-b-[black] pb-4">
              Add Custom Endpoint
            </h3>

            <div className="dark:bg-[#FFFFFF0F] bg-[#FFFFFFCC] p-6 rounded-2xl">
              <Formik
                initialValues={{ rpc: '', api: '', indexer: '', explorer: '' }}
                validationSchema={EndpointSchema}
                onSubmit={async (values, { setSubmitting, resetForm }) => {
                  try {
                    await handleCustomEndpointSubmit(values);
                    resetForm();
                    setIsModalOpen(false);
                  } catch (error) {
                    console.error(error);
                  } finally {
                    setSubmitting(false);
                  }
                }}
              >
                {({ isSubmitting, isValid }) => (
                  <Form className="flex flex-col gap-6">
                    <div className="form-control">
                      <label className="text-sm font-medium mb-2 dark:text-[#FFFFFF99] text-[#000000CC]">
                        RPC Endpoint
                      </label>
                      <div className="relative">
                        <TextInput
                          name="rpc"
                          placeholder="Enter RPC URL"
                          className="w-full px-4 py-3 bg-transparent 
                            dark:bg-[#FFFFFF0F] bg-[#FFFFFFCC]
                            border dark:border-[#FFFFFF20] border-[#00000020]
                            dark:focus:border-[#FFFFFF40] focus:border-[#00000040]
                            rounded-xl transition-all duration-200
                            dark:text-[#FFFFFF99] text-[#000000CC]
                            placeholder:dark:text-[#FFFFFF60] placeholder:text-[#00000060]"
                        />
                      </div>
                    </div>

                    <div className="form-control">
                      <label className="text-sm font-medium mb-2 dark:text-[#FFFFFF99] text-[#000000CC]">
                        API Endpoint
                      </label>
                      <div className="relative">
                        <TextInput
                          name="api"
                          placeholder="Enter API URL"
                          className="w-full px-4 py-3 bg-transparent 
                            dark:bg-[#FFFFFF0F] bg-[#FFFFFFCC]
                            border dark:border-[#FFFFFF20] border-[#00000020]
                            dark:focus:border-[#FFFFFF40] focus:border-[#00000040]
                            rounded-xl transition-all duration-200
                            dark:text-[#FFFFFF99] text-[#000000CC]
                            placeholder:dark:text-[#FFFFFF60] placeholder:text-[#00000060]"
                        />
                      </div>
                    </div>

                    <div className="form-control">
                      <label className="text-sm font-medium mb-2 dark:text-[#FFFFFF99] text-[#000000CC]">
                        Indexer Endpoint
                      </label>
                      <div className="relative">
                        <TextInput
                          name="indexer"
                          placeholder="Enter indexer URL"
                          className="w-full px-4 py-3 bg-transparent
                            dark:bg-[#FFFFFF0F] bg-[#FFFFFFCC]
                            border dark:border-[#FFFFFF20] border-[#00000020]
                            dark:focus:border-[#FFFFFF40] focus:border-[#00000040]
                            rounded-xl transition-all duration-200
                            dark:text-[#FFFFFF99] text-[#000000CC]
                            placeholder:dark:text-[#FFFFFF60] placeholder:text-[#00000060]"
                        />
                      </div>
                    </div>

                    <div className="form-control">
                      <label className="text-sm font-medium mb-2 dark:text-[#FFFFFF99] text-[#000000CC]">
                        Explorer Endpoint
                      </label>
                      <div className="relative">
                        <TextInput
                          name="explorer"
                          placeholder="Enter explorer URL"
                          className="w-full px-4 py-3 bg-transparent
                            dark:bg-[#FFFFFF0F] bg-[#FFFFFFCC]
                            border dark:border-[#FFFFFF20] border-[#00000020]
                            dark:focus:border-[#FFFFFF40] focus:border-[#00000040]
                            rounded-xl transition-all duration-200
                            dark:text-[#FFFFFF99] text-[#000000CC]
                            placeholder:dark:text-[#FFFFFF60] placeholder:text-[#00000060]"
                        />
                      </div>
                    </div>

                    <div className="modal-action flex gap-3 mt-6">
                      <button
                        type="button"
                        onClick={() => setIsModalOpen(false)}
                        className="flex-1 btn focus:outline-none dark:bg-[#FFFFFF0F] bg-[#0000000A] py-2.5 sm:py-3.5"
                      >
                        Cancel
                      </button>
                      <button
                        type="submit"
                        disabled={isSubmitting || !isValid}
                        className="flex-1 btn btn-gradient text-white disabled:text-black 
                          py-2.5 sm:py-3.5 disabled:opacity-50 
                          disabled:cursor-not-allowed transition-all duration-200"
                      >
                        {isSubmitting ? (
                          <span className="flex items-center gap-2">
                            <span className="loading loading-spinner loading-sm"></span>
                            Adding...
                          </span>
                        ) : (
                          'Add Endpoint'
                        )}
                      </button>
                    </div>
                  </Form>
                )}
              </Formik>
            </div>
          </div>
        </div>
      )}
    </dialog>
  );
}

const EndpointSelector = dynamic(() => Promise.resolve(SSREndpointSelector), {
  ssr: false,
});

export default EndpointSelector;
