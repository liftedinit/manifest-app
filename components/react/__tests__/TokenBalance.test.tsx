import { cleanup, render, screen } from '@testing-library/react';
import { afterEach, describe, expect, test } from 'bun:test';

import { TokenBalance } from '@/components';
import { formatComponent } from '@/tests';
import { mockDenomMeta1 } from '@/tests/data';
import { unsafeConvertTokenBase } from '@/utils';

describe('TokenBalance', () => {
  afterEach(cleanup);

  test('should render the token balance', () => {
    const mockup = render(
      <TokenBalance
        token={{
          amount: '1000000000000000000',
          metadata: mockDenomMeta1,
          display: mockDenomMeta1.display,
          base: unsafeConvertTokenBase(mockDenomMeta1.base),
        }}
      />
    );

    expect(screen.getByText('1T')).toBeInTheDocument();
    expect(screen.getByText('TOKEN1')).toBeInTheDocument();
    expect(document.querySelector('.tooltip')).toBeInTheDocument();
    expect(formatComponent(mockup.asFragment())).toMatchSnapshot();
  });

  test('should render the token balance for small amounts', () => {
    const mockup = render(
      <TokenBalance
        token={{
          amount: '1234567',
          metadata: mockDenomMeta1,
          display: mockDenomMeta1.display,
          base: unsafeConvertTokenBase(mockDenomMeta1.base),
        }}
      />
    );

    expect(screen.getByText('1.234567')).toBeInTheDocument();
    expect(screen.getByText('TOKEN1')).toBeInTheDocument();
  });

  test('should not render the tooltip if balance is the same as tooltip', () => {
    const mockup = render(
      <TokenBalance
        token={{
          amount: '1234567',
          metadata: mockDenomMeta1,
          display: mockDenomMeta1.display,
          base: unsafeConvertTokenBase(mockDenomMeta1.base),
        }}
      />
    );

    expect(document.querySelector('.tooltip')).not.toBeInTheDocument();
  });
});
