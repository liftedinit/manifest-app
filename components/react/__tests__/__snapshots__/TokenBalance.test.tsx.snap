// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`TokenBalance should render the token balance 1`] = `
"<DocumentFragment>
  <span
    class="inline-block tooltip token-amount"
    data-tip="1,000,000,000,000 TOKEN1"
  >
    1T 
    <span>
      TOKEN1
    </span>
  </span>
</DocumentFragment>"
`;
